sql/core/src/main/scala/org/apache/spark/sql/execution/systemcatalog/InformationSchema.scala

```diff
@@ -0,0 +1,230 @@
+      .select(requiredColumns.head, requiredColumns.tail: _*).rdd
+  }
+}
+
+class ViewsSource extends SchemaRelationProvider {
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String],
+      schema: StructType): BaseRelation = {
+    ViewsRelation(sqlContext.sparkSession)
+  }
+}
+
+case class ViewsRelation(@transient sparkSession: SparkSession)
+  extends BaseRelation with PrunedFilteredScan {
+
+  override def sqlContext: SQLContext = sparkSession.sqlContext
+
+  override def schema: StructType = StructType(Seq(
+    StructField("TABLE_CATALOG", StringType, false),
+    StructField("TABLE_SCHEMA", StringType, false),
+    StructField("TABLE_NAME", StringType, false),
+    StructField("VIEW_DEFINITION", StringType, false)
+  ))
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    sparkSession.catalog.listTables()
+      .where("isTemporary")
+      .selectExpr(
+        "'default' as TABLE_CATALOG",
+        "IFNULL(database, 'default') as TABLE_SCHEMA",
+        "name as TABLE_NAME",
+        "IFNULL(description,'VIEW') as VIEW_DEFINITION") // TODO
+      .select(requiredColumns.head, requiredColumns.tail: _*).rdd
+  }
+}
+
+class ColumnsSource extends SchemaRelationProvider {
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String],
+      schema: StructType): BaseRelation = {
+    ColumnsRelation(sqlContext.sparkSession)
+  }
+}
+
+case class ColumnsRelation(@transient sparkSession: SparkSession)
+  extends BaseRelation with PrunedFilteredScan {
+
+  override def sqlContext: SQLContext = sparkSession.sqlContext
+
+  override def schema: StructType = StructType(Seq(
+    StructField("TABLE_CATALOG", StringType, false),
+    StructField("TABLE_SCHEMA", StringType, false),
+    StructField("TABLE_NAME", StringType, false),
+    StructField("COLUMN_NAME", StringType, false),
+    StructField("ORDINAL_POSITION", LongType, false),
+    StructField("IS_NULLABLE", BooleanType, false),
+    StructField("DATA_TYPE", StringType, false)
+  ))
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val tables = sparkSession.catalog.listTables().collect()
+    val result = new ArrayBuffer[Row]
+    for (t <- tables) {
+      val database = t.database
+      val table = t.name
+      if (t.database == null) {
+        // TODO: SPARK-16458
+      } else {
+        result ++= sparkSession.catalog.listColumns(t.database, t.name).rdd.zipWithIndex.map {
+          case (col, index) =>
+            Row("default", database, table, col.name, index, col.nullable, col.dataType)
+        }.collect()
+      }
+    }
+    sparkSession
+      .createDataFrame(sparkSession.sparkContext.parallelize(result), schema)
+      .select(requiredColumns.head, requiredColumns.tail: _*).rdd
+  }
+}
+
+class SessionVariablesSource extends SchemaRelationProvider {
```

Comment:
can we rename this SessionVariablesRelationProvider and then inline SessionVariablesRelation in here?

e.g.

```
def createRelation(...) = {
  new BaseRelation with PrunedFilteredScan {

  }
}
```